contracts/index/bases/IndexStrategyUpgradeable.sol

```diff
@@ -0,0 +1,618 @@
+
+    modifier onlyWhitelistedToken(address token) {
+        if (!isTokenWhitelisted(token)) {
+            revert Errors.Index_NotWhitelistedToken(token);
+        }
+
+        _;
+    }
+
+    modifier whenNotReachedEquityValuationLimit() {
+        _;
+
+        if (equityValuation(true, true) > equityValuationLimit) {
+            revert Errors.Index_ExceedEquityValuationLimit();
+        }
+    }
+
+    // solhint-disable-next-line
+    function __IndexStrategyUpgradeable_init(
+        IndexStrategyInitParams calldata initParams
+    ) internal onlyInitializing {
+        __ERC165_init();
+        __ReentrancyGuard_init();
+        __Context_init();
+        __Ownable_init();
+        __Pausable_init();
+
+        wETH = initParams.wETH;
+
+        indexToken = IIndexToken(initParams.indexToken);
+
+        for (uint256 i = 0; i < initParams.components.length; i++) {
+            components.push(initParams.components[i].token);
+
+            _setWeight(
+                initParams.components[i].token,
+                initParams.components[i].weight
+            );
+        }
+
+        for (uint256 i = 0; i < initParams.swapRoutes.length; i++) {
+            setSwapRoute(
+                initParams.swapRoutes[i].token0,
+                initParams.swapRoutes[i].token1,
+                initParams.swapRoutes[i].router,
+                initParams.swapRoutes[i].dex,
+                initParams.swapRoutes[i].pairData
+            );
+        }
+
+        setWhitelistedTokens(initParams.whitelistedTokens);
+
+        setOracle(initParams.oracle);
+
+        setEquityValuationLimit(initParams.equityValuationLimit);
+    }
+
+    function pause() external onlyOwner {
+        super._pause();
+    }
+
+    function unpause() external onlyOwner {
+        super._unpause();
+    }
+
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override
+        returns (bool)
+    {
+        return
+            interfaceId == type(IIndexStrategy).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    function mintExactIndexFromToken(
+        address token,
+        uint256 amountTokenMax,
+        uint256 amountIndex,
+        address recipient
+    )
+        external
+        nonReentrant
+        whenNotPaused
+        onlyWhitelistedToken(token)
+        whenNotReachedEquityValuationLimit
+        returns (uint256 amountTokenSpent)
+    {
+        if (recipient == address(0)) {
+            revert Errors.Index_ZeroAddress();
+        }
+
+        IERC20Upgradeable(token).safeTransferFrom(
+            _msgSender(),
+            address(this),
+            amountTokenMax
+        );
+
+        (
+            uint256 amountWETHTotal,
+            uint256[] memory amountWETHs,
+            address[] memory bestRouters,
+            uint256[] memory amountComponents
+        ) = _getMintingDataForExactIndex(amountIndex);
+
+        if (token != wETH) {
```

Comment:
let's say token input parameter is TokenA
and one of the index component is TokenB
why are we doing a TokenA->WETH->TokenB swap here? What is the most efficient swap path is TokenA->TokenB? Why are we constratining ourselves to swap through WETH